Fix changing locus from SingleQE to OuterQuery

red1452 opened the pull request:
Fix changing locus from SingleQE to OuterQuery

cdbpath_create_motion_path() has optimization, which allows not to add extra
motion node for cases, when locus is changed from SingleQE to OuterQuery.
cdbpath_create_motion_path() just changes locus at the first node in the path.
It is correct when path contains just one Motion node. But Outer locus requires
Materialization. The Materialize node must be added before the Motion node.

It is actual for queries with subquery containing both of volatile function and
replicated table. handle_gen_seggen_volatile_path() changes locus to SingleQE
and adds Motion and ProjectionPath node. Next, bring_to_outer_query() changes
locus to OuterQuery and add ProjectionPath with outer quals.

This patch modifies the cdbpath_create_motion_path function for cases when
SingleQE locus is changed to the Outer locus. If the Motion node is the first
node in the path, locus will be changed to Outer and Materialize node will be
added. If there is a Projection path with Motion below, their locus will be
updated, and a Materialize path will be added in between of them. For other
cases, Materialize and Motion paths will just be added on top without modifying
other nodes.

The problem was found when tests from commit
7ef4218221babf046b63f748d3467b54ba5c9ca0 were run. These test are usefull,
because it check correctness of planning queries with volatile functions and
replicated tables. But  another code changes can not be taken because of
difference between 6X and 7X. These tests were added and expanded by adding
execution of queries, which do not contain random().

Stolb27 commented on src/backend/cdb/cdbllize.c:
```diff
@@ -798,6 +798,21 @@ cdbllize_decorate_subplans_with_motions(PlannerInfo *root, Plan *plan)
 			subplan->flow->locustype != CdbLocusType_General)
 		{
+			/*
+			 * If the first node is "Result", adding "Materialize" and "Motion"
```
Do you have any guaranties that Result node always top-most in the subplan?

Stolb27 replied:
Do you have any guaranties that Result node always top-most in the subplan?

Stolb27 commented on src/test/regress/expected/rpt.out:
```diff
@@ -1608,6 +1658,182 @@ drop table t1;
+         Filter: (SubPlan 1)
+         SubPlan 1
+           ->  Result
+                 Output: f(t.i)
```
This plan seems to be still invalid, because the executor may calculate different value of volatile function on each segment for distribution, declared as Replicated.

red1452 replied:
I checked different queries on GP and Postgres and got the same resullts on each BD.
For example one of such queries.
GP:
```
postgres=# explain (costs off, verbose, analyze) select * from t1 where a in (select f(i) from t where i=a and f(i) > 0 group by i order by i limit 100);
                                                          QUERY PLAN                                                          
------------------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3) (actual time=1.687..1.756 rows=10 loops=1)
   Output: t1.a
   ->  Seq Scan on public.t1 (actual time=1.222..1.316 rows=5 loops=1)
         Output: t1.a
         Filter: (SubPlan 1)
         SubPlan 1
           ->  Limit (actual time=0.040..0.056 rows=1 loops=5)
                 Output: (f(t.i)), t.i
                 ->  GroupAggregate (actual time=0.040..0.055 rows=1 loops=5)
                       Output: f(t.i), t.i
                       Group Key: t.i
                       ->  Result (actual time=0.001..0.005 rows=1 loops=5)
                             Output: t.i
                             Filter: (t.i = t1.a)
                             ->  Materialize (actual time=0.162..0.163 rows=8 loops=6)
                                   Output: t.i
                                   work_mem: 50kB  Segments: 3  Max: 17kB (segment 0)  Workfile: (0 spilling)
                                   ->  Broadcast Motion 1:3  (slice2; segments: 1) (actual time=0.961..0.964 rows=10 loops=1)
                                         Output: t.i
                                         ->  Seq Scan on public.t (actual time=0.245..0.304 rows=10 loops=1)
                                               Output: t.i
                                               Filter: (f(t.i) > 0)
                                               Rows Removed by Filter: 1
 Optimizer: Postgres-based planner
 Planning Time: 2.474 ms
   (slice0)    Executor memory: 57K bytes.
   (slice1)    Executor memory: 77K bytes avg x 3 workers, 77K bytes max (seg0).  Work_mem: 17K bytes max.
   (slice2)    Executor memory: 46K bytes (seg2).
 Memory used:  128000kB
 Execution Time: 5.760 ms
(30 rows)

postgres=# select * from t1 where a in (select f(i) from t where i=a and f(i) > 0 group by i order by i limit 100);
 a  
----
  1
  5
  6
  9
 10
  2
  3
  4
  7
  8
(10 rows)
```
Postgres:
```
test=# explain (costs off, verbose, analyze) select * from t1 where a in (select f(i) from t where i=a and f(i) > 0 group by i order by f(i) limit 100);
                                        QUERY PLAN                                         
-------------------------------------------------------------------------------------------
 Seq Scan on public.t1 (actual time=61.550..61.595 rows=10 loops=1)
   Output: t1.a
   Filter: (SubPlan 1)
   Rows Removed by Filter: 1
   SubPlan 1
     ->  Limit (actual time=0.012..0.012 rows=1 loops=11)
           Output: (f(t.i)), t.i
           ->  Sort (actual time=0.011..0.011 rows=1 loops=11)
                 Output: (f(t.i)), t.i
                 Sort Key: (f(t.i))
                 Sort Method: quicksort  Memory: 25kB
                 ->  Group (actual time=0.010..0.010 rows=1 loops=11)
                       Output: f(t.i), t.i
                       Group Key: t.i
                       ->  Seq Scan on public.t (actual time=0.007..0.007 rows=1 loops=11)
                             Output: t.i
                             Filter: ((t.i = t1.a) AND (f(t.i) > 0))
                             Rows Removed by Filter: 10
 Query Identifier: 2321456564250273903
 Planning Time: 0.187 ms
 Execution Time: 63.126 ms
(21 rows)

select * from t1 where a in (select f(i) from t where i=a and f(i) > 0 group by i order by f(i) limit 100);
 a  
----
  1
  2
  3
  4
  5
  6
  7
  8
  9
 10
(10 rows)
```

Stolb27 replied:
> I checked different queries on GP and Postgres and got the same resullts on each BD.

Equivalence of results doesn't proof that distributed plan is valid in the common case.

red1452 replied:
I generated table with the same values at different segments and created function, which returns the same value with 0.5 probability:
```sql
create table t (i int) distributed replicated;
create table t3 (a int, b int) distributed by (b);
insert into t select generate_series(0,10);
insert into t3 select i, i+20 from generate_series(0,10)i;
insert into t3 select i, i+30 from generate_series(0,10)i;

create or replace function f2(i int)
returns int
language sql security definer
as $$
	select case when random() > 0.5 then i + 1
	       else i end;
$$;
```
```
postgres=# select *,gp_segment_id from t3 where a = 2; 
 a  | b  | gp_segment_id 
----+----+---------------
  2 | 22 |             0
  2 | 32 |             2
(2 rows)
```
I think, that query should return result with both rows or neither, but I got different results:
```
postgres=# select *,gp_segment_id from t3 where a in (select f2(i) from t where i=a and i = 2 and f2(i) > 0);
 a | b | gp_segment_id 
---+---+---------------
(0 rows)

postgres=# select *,gp_segment_id from t3 where a in (select f2(i) from t where i=a and i = 2 and f2(i) > 0);
 a | b  | gp_segment_id 
---+----+---------------
 2 | 32 |             2
(1 row)

postgres=# select *,gp_segment_id from t3 where a in (select f2(i) from t where i=a and i = 2 and f2(i) > 0);
 a | b  | gp_segment_id 
---+----+---------------
 2 | 22 |             0
 2 | 32 |             2
(2 rows)

postgres=# select *,gp_segment_id from t3 where a in (select f2(i) from t where i=a and i = 2 and f2(i) > 0);
 a | b  | gp_segment_id 
---+----+---------------
 2 | 22 |             0
(1 row)

postgres=# select *,gp_segment_id from t3 where a in (select f2(i) from t where i=a and i = 2 and f2(i) > 0);
 a | b  | gp_segment_id 
---+----+---------------
 2 | 32 |             2
(1 row)

postgres=# select *,gp_segment_id from t3 where a in (select f2(i) from t where i=a and i = 2 and f2(i) > 0);
 a | b | gp_segment_id 
---+---+---------------
(0 rows)
```
The same results were gotten at 6X and Postgres. For Postgres I made small changes:
```sql
create table t (i int);
create table t3 (a int, b int);
insert into t select generate_series(0,10);
insert into t3 select i, i+20 from generate_series(0,10)i;
insert into t3 select i, i+30 from generate_series(0,10)i;

create or replace function f2(i int)
returns int
language sql security definer
as $$
	select case when random() > 0.5 then i + 1
	       else i end;
$$;
```
Postgres has such plan:
```
postgres=# explain (costs off, verbose, analyze) select * from t3 where a in (select f2(i) from t where i=a and i = 2);
                                  QUERY PLAN                                  
------------------------------------------------------------------------------
 Seq Scan on public.t3 (actual time=0.177..0.229 rows=2 loops=1)
   Output: t3.a, t3.b
   Filter: (SubPlan 1)
   Rows Removed by Filter: 20
   SubPlan 1
     ->  Result (actual time=0.008..0.008 rows=0 loops=22)
           Output: f2(t.i)
           One-Time Filter: (t3.a = 2)
           ->  Seq Scan on public.t (actual time=0.005..0.005 rows=1 loops=2)
                 Output: t.i
                 Filter: (t.i = 2)
                 Rows Removed by Filter: 1
 Query Identifier: -6850848553024840045
 Planning Time: 0.151 ms
 Execution Time: 0.270 ms
(15 rows)
```
If I understand correctly, Postgres remakes ` SubPlan 1` for each rows at `t3` and it may get different result, because function `f2(int)` contains `random()`. In case of Greenplum it means, that it does not matter, is there the same values at differrent segments or not, because subplan is remade for each row.

Stolb27 replied:
> The same results were gotten at 6X and Postgres.

Equivalence of results doesn't proof that distributed plan is valid in the common case.

> it may get different result, because function f2(int) contains random()

It's a special case. In generic, we mustn't make any hypothesis about its return value.
With constant predicate `= 2` last example relies only on presence of appropriate lines.

red1452 replied:
In the common case nodes `seq scan` -> `motion` -> `materialize` generates equal rows for each segments. After it, applying filter and returning result of volatile function may change different result on different segments, but it is normal. The main idea - is that the input rows are equal for each segments. It is guaranteed by scanning replicated table at one segment (with applying any filters), than result is sent to all segments. All next nodes can apply changes, which are led to different results on different segments and it is expected result, because we use volatile function.

red1452 committed:
Remove changing volatile to singleqe

Stolb27 commented on src/test/regress/expected/rpt_optimizer.out:
```diff
@@ -953,15 +953,81 @@ ___________
+----------------------------------------------------
+ Delete on t_replicate_volatile
+   ->  Seq Scan on t_replicate_volatile
+         Filter: ((a)::double precision < random())
```
Are you sure that this plan is correct? Do we have any warranties that the same tuples will be deleted on different segments?

Stolb27 replied:
It seems that there is no way to perform such queries.

red1452 committed:
partial revert

red1452 committed:
remove useless changes

red1452 committed:
Merge branch 'adb-7.2.0' into ADBDEV-5518

red1452 committed:
Merge branch 'adb-7.2.0' into ADBDEV-5518

Stolb27 commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -754,8 +754,9 @@ set_rel_pathlist(PlannerInfo *root, RelOptInfo *rel,
+	 * Do not do it, if there is outer query.
 	 */
-	if (rel->reloptkind == RELOPT_BASEREL)
+	if (rel->upperrestrictinfo == NULL && rel->reloptkind == RELOPT_BASEREL)
```
Why does empty upperrestrictinfo sign that there is no outer queries?

red1452 replied:
I updated the comment above.

Stolb27 replied:
I see. You filtered out cases for bring_to_outer_query from this condition. Why problem manifests itself only in case of not null `upperrestrictinfo`? What are other cases to execute Segment General only on Single QE?

> Function handle_gen_seggen_volatile_path was not removed, because it contains
check, which restricts DML queries on replicated tables with filter, which
contains volatile functions.

It's quite strange to leave complex transformation logic only for exact checks. Do we still needed these transformations?

red1452 replied:
Functions `handle_gen_seggen_volatile_path`, `bring_to_outer_query` and `bring_to_singleQE` make the same things, but in the different cases. Function `handle_gen_seggen_volatile_path` works when locus is general or segment general or replicated and there is volatile function. Function `bring_to_outer_query` works when locus is not general and not outer query and there is outer quals. Function `bring_to_singleQE` works when locus is not general and not entry and not single QE and not outer query.
I think removing or union of these functions must be done in the other ticket, which will be complex refactoring of it. My patch just fix double work of these functions, when `handle_gen_seggen_volatile_path` and `bring_to_outer_query` work together. It does not needed to execute function `handle_gen_seggen_volatile_path` if function `bring_to_outer_query` will be executed (in case of non null `upperrestrictinfo`).

red1452 replied:
One of some queries, which will be changed if `handle_gen_seggen_volatile_path` is removed:
```
create table t_hashdist(a int, b int, c int) distributed by (a);
create table t_replicate_volatile(a int, b int, c int) distributed replicated;
explain (costs off) select * from t_replicate_volatile, t_hashdist where t_replicate_volatile.a > random();
```
Plan before:
```
                              QUERY PLAN                              
----------------------------------------------------------------------
 Nested Loop
   ->  Gather Motion 3:1  (slice1; segments: 3)
         ->  Seq Scan on t_hashdist
   ->  Materialize
         ->  Result
               ->  Gather Motion 1:1  (slice2; segments: 1)
                     ->  Seq Scan on t_replicate_volatile
                           Filter: ((a)::double precision > random())
 Optimizer: Postgres query optimizer
(9 rows)
```
Plan after:
```
                        QUERY PLAN                        
----------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   ->  Nested Loop
         ->  Seq Scan on t_replicate_volatile
               Filter: ((a)::double precision > random())
         ->  Materialize
               ->  Seq Scan on t_hashdist
 Optimizer: Postgres-based planner
(7 rows)
```

RekGRpth replied:
Could you update the branch and provide an example with data, because I can't get this plan to work with the patch?

red1452 replied:
Simple example from the ticket. Without patch:
```
postgres=# create table t (i int) distributed replicated;
create table t1 (a int) distributed by (a);
CREATE TABLE
CREATE TABLE
postgres=# create or replace function f(i int) returns int language sql security definer as $$ select i; $$;
CREATE FUNCTION
postgres=# insert into t select generate_series(0,10);
insert into t1 select generate_series(0,10);
INSERT 0 11
INSERT 0 11
postgres=# select * from t1 where a in (select f(i) from t where i=a and f(i) > 0);
ERROR:  Passing parameters across motion is not supported. (cdbmutate.c:2051)
postgres=# explain (costs off, verbose) select * from t1 where a in (select f(i) from t where i=a and f(i) > 0);
ERROR:  Passing parameters across motion is not supported. (cdbmutate.c:2051)
```
The same query with patch:
```
postgres=# create table t (i int) distributed replicated;
create table t1 (a int) distributed by (a);
CREATE TABLE
CREATE TABLE
postgres=# create or replace function f(i int) returns int language sql security definer as $$ select i; $$;
CREATE FUNCTION
postgres=# insert into t select generate_series(0,10);
insert into t1 select generate_series(0,10);
INSERT 0 11
INSERT 0 11
postgres=# select * from t1 where a in (select f(i) from t where i=a and f(i) > 0);
 a  
----
  1
  2
  3
  4
  7
  8
  5
  6
  9
 10
(10 rows)

postgres=# explain (costs off, verbose) select * from t1 where a in (select f(i) from t where i=a and f(i) > 0);
                              QUERY PLAN                               
-----------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   Output: t1.a
   ->  Seq Scan on public.t1
         Output: t1.a
         Filter: (SubPlan 1)
         SubPlan 1
           ->  Result
                 Output: f(t.i)
                 Filter: (t.i = t1.a)
                 ->  Materialize
                       Output: t.i
                       ->  Broadcast Motion 1:3  (slice2; segments: 1)
                             Output: t.i
                             ->  Seq Scan on public.t
                                   Output: t.i
                                   Filter: (f(t.i) > 0)
 Optimizer: Postgres-based planner
(17 rows)
```

Stolb27 reviewed (commented):
Please reword PR description. Why adding of Motion by the mentioned function is incorrect? Which consequences is leaded? The `bring_to_outer_query` is able to produce both kind of motions (Gather or Broadcast). What is the case when the `handle_gen_seggen_volatile_path` still needed that makes you to exclude it only partially?

red1452 committed:
update comment

Stolb27 commented:
> Function handle_gen_seggen_volatile_path is used in 6X to make single QE query
from queries on replicated tables with volatile function. It adds Motion and
Result nodes to make query at one segment correctly.

In my opinion, the function transforms only part of query, not the whole one.

red1452 committed:
Merge branch 'adb-7.2.0' into ADBDEV-5518

red1452 committed:
Merge branch 'adb-7.2.0' into ADBDEV-5518

red1452 committed:
Merge branch 'adb-7.2.0' into ADBDEV-5518

Stolb27 commented:
> Remove changing volatile to singleqe for queries with outer quals

How are connected volatile type of function with data location (locus) to be converted one to another?

> Function handle_gen_seggen_volatile_path is used in 6X to make single QE from
part of query which contains replicated table and volatile function. It adds
Motion and Result nodes to make such part of query at one segment correctly. In
7X it does not need to do, because if query contains outer quals, nodes result
and materialize will be added by bring_to_outer_query.

In my opinion, these functions solve different problems and have not to conflict between them. The 1st one handles volatile predicates and segregate them to another slice, the 2nd one makes sure that outer quals are applied in the outer motion locus. So, exclusion one of them seems to be a weird solution.

red1452 committed:
Merge branch 'adb-7.2.0' into ADBDEV-5518

red1452 committed:
fix expected output of ORCA's test

red1452 commented:
> In my opinion, these functions solve different problems and have not to conflict between them. The 1st one handles volatile predicates and segregate them to another slice, the 2nd one makes sure that outer quals are applied in the outer motion locus. So, exclusion one of them seems to be a weird solution.

These functions may intersects. The first function depends on locus and volatile function, the second function depends on existing outer quals. If both of functions apply changes, we will get two Motion nodes (one of these motions will be extra). And quals will be be under extra motion. It will lead to incorrect result, because Motion node will not be re-scaned.

Stolb27 commented:
> If both of functions apply changes, we will get two Motion nodes (one of these motions will be extra). And quals will be be under extra motion.

The 2nd function shouldn't add motion because [considers](https://github.com/arenadata/gpdb/blob/2a6c60d12d7ac37ed843a08dc3d824e96af3b93f/src/backend/cdb/cdbpath.c#L218-L245) outcome distribution of the 1st one as compatible.

red1452 commented:
> bring_to_outer_query

> The 2nd function shouldn't add motion because [considers](https://github.com/arenadata/gpdb/blob/2a6c60d12d7ac37ed843a08dc3d824e96af3b93f/src/backend/cdb/cdbpath.c#L218-L245) outcome distribution of the 1st one as compatible.

The first function changes locus to singleQE (and adds Motion node), The second function just changes locus from  singleQE to outerQuery:
```
Thread 1 "postgres" hit Breakpoint 7, cdbpath_create_motion_path (root=0x5c8317250970, subpath=0x5c83172522c0, pathkeys=0x0, require_existing_order=false, locus=...) at cdbpath.c:244
244                             return subpath;
(gdb) bt
#0  cdbpath_create_motion_path (root=0x5c8317250970, subpath=0x5c83172522c0, pathkeys=0x0, require_existing_order=false, locus=...) at cdbpath.c:244
#1  0x00005c830f646a19 in bring_to_outer_query (root=0x5c8317250970, rel=0x5c83171c8b50, outer_quals=0x5c8317253070) at allpaths.c:551
#2  0x00005c830f646f77 in set_rel_pathlist (root=0x5c8317250970, rel=0x5c83171c8b50, rti=1, rte=0x5c831724ffd8) at allpaths.c:771
#3  0x00005c830f6465b0 in set_base_rel_pathlists (root=0x5c8317250970) at allpaths.c:377
#4  0x00005c830f6462df in make_one_rel (root=0x5c8317250970, joinlist=0x5c8317252bb8) at allpaths.c:244
#5  0x00005c830f685721 in query_planner (root=0x5c8317250970, qp_callback=0x5c830f68d161 <standard_qp_callback>, qp_extra=0x7ffe7f88b3c0) at planmain.c:303
#6  0x00005c830f689b48 in grouping_planner (root=0x5c8317250970, inheritance_update=false, tuple_fraction=0.5) at planner.c:2482
#7  0x00005c830f687951 in subquery_planner (glob=0x5c831724d920, parse=0x5c831724fe60, parent_root=0x5c831724f3e0, hasRecursion=false, tuple_fraction=0.5, config=0x5c8317250950) at planner.c:1280
#8  0x00005c830f69c5b1 in make_subplan (root=0x5c831724f3e0, orig_subquery=0x5c83171df068, subLinkType=ANY_SUBLINK, subLinkId=0, testexpr=0x5c831724fcf0, isTopQual=true) at subselect.c:407
(gdb) p subpath->locus
$3 = {locustype = CdbLocusType_OuterQuery, distkey = 0x0, numsegments = -1}
```
The second Motion is added by cdbllize_adjust_init_plan_path:
```
Thread 1 "postgres" hit Breakpoint 3, cdbpath_create_motion_path (root=0x5c831724f3e0, subpath=0x5c8317253300, pathkeys=0x0, require_existing_order=false, locus=...) at cdbpath.c:321
321                             return (Path *) pathnode;
(gdb) bt
#0  cdbpath_create_motion_path (root=0x5c831724f3e0, subpath=0x5c8317253300, pathkeys=0x0, require_existing_order=false, locus=...) at cdbpath.c:321
#1  0x00005c830fac59c0 in cdbllize_adjust_init_plan_path (root=0x5c831724f3e0, best_path=0x5c8317253300) at cdbllize.c:673
#2  0x00005c830f69c717 in make_subplan (root=0x5c831724f3e0, orig_subquery=0x5c83171df068, subLinkType=ANY_SUBLINK, subLinkId=0, testexpr=0x5c831724fcf0, isTopQual=true) at subselect.c:447
```

Interesting case: if remove bring_to_outer_query, parametrized subplan will be simple subplan without params and will work correctly:
```
postgres=# explain (costs off, verbose, analyze) select * from t1 where a in (select f(i) from t where i=a and f(i) > 0);
                                                     QUERY PLAN                                                      
---------------------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3) (actual time=4.453..4.601 rows=10 loops=1)
   Output: t1.a
   ->  Seq Scan on public.t1 (actual time=3.673..3.706 rows=5 loops=1)
         Output: t1.a
         Filter: (SubPlan 1)
         SubPlan 1
           ->  Materialize (actual time=0.730..0.733 rows=6 loops=5)
                 Output: (f(t.i))
                 work_mem: 50kB  Segments: 3  Max: 17kB (segment 0)  Workfile: (0 spilling)
                 ->  Broadcast Motion 1:3  (slice2) (actual time=3.540..3.551 rows=10 loops=1)
                       Output: (f(t.i))
                       ->  Result (actual time=2.942..3.254 rows=10 loops=1)
                             Output: f(t.i)
                             ->  Gather Motion 1:1  (slice3; segments: 1) (actual time=2.497..2.510 rows=10 loops=1)
                                   Output: t.i
                                   ->  Seq Scan on public.t (actual time=0.772..1.098 rows=10 loops=1)
                                         Output: t.i
                                         Filter: (f(t.i) > 0)
                                         Rows Removed by Filter: 1
 Optimizer: Postgres-based planner
 Planning Time: 1.419 ms
   (slice0)    Executor memory: 48K bytes.
   (slice1)    Executor memory: 40K bytes avg x 3 workers, 40K bytes max (seg0).  Work_mem: 17K bytes max.
   (slice2)    Executor memory: 45K bytes (entry db).
   (slice3)    Executor memory: 45K bytes (seg2).
 Memory used:  128000kB
 Execution Time: 6.276 ms
(27 rows)
```

red1452 committed:
major changes, move fix to bring_to_outer_query

red1452 committed:
fix and expand tests

Stolb27 commented on src/test/regress/expected/rpt.out:
```diff
@@ -1613,6 +1663,210 @@ drop table t1;
+
+-- ensure we make broadcast motion when volatile function in correlated subplan qual
+explain (costs off, verbose) select * from t1 where a in (select f(i) from t where i=a and f(i) > 0);
+                                QUERY PLAN                                
```
This plan doesn't check correlated predicate and as a result may produce invalid output.

red1452 committed:
Revert "major changes, move fix to bring_to_outer_query"

This reverts commit edf4ac96ab4bc51c41f127297a37dba20baec841.

red1452 committed:
fix test

red1452 committed:
fix cdbpath_create_motion_path for outer query]

red1452 committed:
remove extra changes

red1452 committed:
Fix changing locus from SingleQE to OuterQuery at cdbpath_create_motion_path

Function cdbpath_create_motion_path just changes locus from General, Entry or
SingleQE locus to OuterQuery locus at the first node in path. It may incorrect
when path has already contained the Motion node or Motion node after
ProjectionPath. At such cases The Motion must be protected from by adding
Materialize node before Motion. And locus should be changed at the Motion node.

Such cases may be cased, when subquery contains both of volatile function and
replicated table. Function handle_gen_seggen_volatile_path changes locus to
SingleQE and adds Motion and ProjectionPath node. Next, bring_to_outer_query
changes locus to OuterQuery and add ProjectionPath with outer quals.

Also, this patch adds tests from commit 7ef4218221babf046b63f748d3467b54ba5c9ca0
and expand these tests by adding execution of queries, which do not contain
random().

Stolb27 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -224,24 +224,36 @@ cdbpath_create_motion_path(PlannerInfo *root,
+				return (Path *) create_material_path(root, subpath->parent,
+													 subpath);
+			}
+			else if (IsA(subpath, ProjectionPath))
```
Why only these nodes are considered to check?

red1452 replied:
These nodes are checked, because these nodes are added to correctly change locus. If these nodes are presented, we can change it without adding new the same nodes.

Stolb27 replied:
How will be handled other kinds of paths? Remember, this function is used widely.

red1452 replied:
For other kinds of path the Materialize and Motion nodes will be added to the top at the and of the function.

Stolb27 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -224,24 +224,36 @@ cdbpath_create_motion_path(PlannerInfo *root,
+					IsA(projection_path->subpath, CdbMotionPath))
+				{
+					subpath = projection_path->subpath;
+					subpath->locus = locus;
```
You alter only Motion path locus and leave Projection path locus as is. Is it correct in the common case?

red1452 replied:
I added changing locus at Projection path.

Stolb27 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -224,24 +224,36 @@ cdbpath_create_motion_path(PlannerInfo *root,
-		}
-
-		if (CdbPathLocus_IsEntry(subpath->locus) ||
+		if (CdbPathLocus_IsGeneral(subpath->locus) ||
+			CdbPathLocus_IsEntry(subpath->locus) ||
```
1. May the path of General locus contain motions?
2. Is it valid to alter Entry/SingleQE data flow to Outer always? They will be able to execute on multiple segments as Outer locus allows.

red1452 replied:
My patch does this for cases, when there is just one Motion node or Projection path and Motion node (and it also adds Materialize node, as Outer locus requires). For other cases new Materialize and Motion nodes are added at the end of function.

red1452 committed:
Merge remote-tracking branch 'origin/adb-7.2.0' into ADBDEV-5518

red1452 committed:
add comment and check count of segments

Stolb27 commented on src/backend/cdb/cdbpath.c:
```diff
@@ -235,13 +235,45 @@ cdbpath_create_motion_path(PlannerInfo *root,
-			 *
-			 * This is also bogus, because the outer query might need to run
-			 * in segments.
+			 * If count of segments is not changed, just change locus.
```
Why it's valid to alter only locus for cases with equal number of segments? Even in case of result with motion below?

Stolb27 reviewed (commented):
Function cdbpath_create_motion_path just changes locus from **General**, **Entry** or SingleQE locus to OuterQuery locus at the first node in path. 

> PR description have to be updated

red1452 committed:
remove useless comparing

red1452 committed:
Merge branch 'adb-7.2.0' into ADBDEV-5518

red1452 committed:
revert changes for entry locus

red1452 committed:
fix comment

Stolb27 commented on src/test/regress/expected/rpt_optimizer.out:
```diff
@@ -955,6 +955,58 @@ ___________
+                                 Sort Key: t_replicate_src.i
+                                 ->  Seq Scan on rpt.t_replicate_src
+                                       Output: t_replicate_src.i
+ Optimizer: Postgres-based planner
```
This fallback is a known issue? What about other fallbacks?

```
INFO:  GPORCA failed to produce a plan, falling back to Postgres-based planner
DETAIL:  CDistributionSpecReplicated.cpp:135: Failed assertion: Edt() != CDistributionSpec::EdtTaintedReplicated
Stack trace:
1    0x00005d91c11945e8 gpos::CErrorContext::Record + 88
2    0x00005d91c119530f gpos::CException::Raise + 255
3    0x00005d91c1393779 gpopt::CDistributionSpecReplicated::AppendEnforcers + 1001
```

red1452 replied:
I created new [ticket](https://tracker.yandex.ru/ADBDEV-7517) for this fallback.
The other fallbacks is normal case with correct error message:
```
Falling back to Postgres-based planner because GPORCA does not support the following feature: SIRV functions
```

Stolb27 commented on src/test/regress/expected/rpt.out:
```diff
@@ -1613,6 +1663,209 @@ drop table t1;
+    10
+(1 row)
+
+-- ensure we do not make broadcast motion
```
Plan below contains broadcast motion. What is this test description meant?

Is this plan below valid? Why two plans below looks similar? Where is grouping? Is it really redundant?

red1452 replied:
I removed this query from test and created new [ticket](https://tracker.yandex.ru/ADBDEV-7518).

Stolb27 reviewed (commented):
As a conclusion, the given patch fix changing locus type for known SingleQE Motions cases. These cases may be considered as any motion kind is appliable, but there are no fitting locus type to express it. Other transformations from SingleQE to Outer require motion now.

red1452 committed:
fix test

red1452 committed:
Merge branch 'adb-7.2.0' into ADBDEV-5518

bandetto commented on src/backend/cdb/cdbpath.c:
```diff
@@ -243,6 +242,47 @@ cdbpath_create_motion_path(PlannerInfo *root,
+			 * change locus to Outer. Also there may be projection path node,
+			 * just change locus at the such node, too. For other cases Motion
+			 * and Materialize nodes will added at the end of this function.
+			 */
```
This comment just explains the what below code does, but does not exactly explain why that logic is needed. How about replacing this comment with an explanation why?

bandetto commented on src/backend/cdb/cdbpath.c:
```diff
@@ -243,6 +242,47 @@ cdbpath_create_motion_path(PlannerInfo *root,
+				ProjectionPath* projection_path = (ProjectionPath *) subpath;
+
+				if (projection_path->subpath &&
+					IsA(projection_path->subpath, CdbMotionPath))
```
Perhaps use `PointerIsValid(...)` or `... != NULL` to avoid confusing the field with a boolean condition?
```suggestion
				if (PointerIsValid(projection_path->subpath) &&
					IsA(projection_path->subpath, CdbMotionPath))
```

bandetto commented on src/backend/cdb/cdbpath.c:
```diff
@@ -243,6 +242,47 @@ cdbpath_create_motion_path(PlannerInfo *root,
+					projection_path->subpath = (Path *) create_material_path(root,
+																subpath->parent,
+																subpath);
+					return (Path *) projection_path;
```
Maybe avoid reassigning to `subpath` to help with readability?
```suggestion
					subpath->locus = locus;
					projection_path->subpath->locus = locus;
					projection_path->subpath = (Path *) create_material_path(
						root, projection_path->subpath->parent,
						projection_path->subpath);
					return (Path *) projection_path;
```

bandetto commented on src/backend/cdb/cdbpath.c:
```diff
@@ -243,6 +242,47 @@ cdbpath_create_motion_path(PlannerInfo *root,
+			/*
+			 * Just one Motion node. Change locus to Outer and add Materialize.
+			 */
+			if (IsA(subpath, CdbMotionPath))
```
Looking at the node tag here seems like a hack to me. Is this the only way to check for the aforementioned case? Can the logic be moved to functions which added these nodes initially?

red1452 replied:
It is not a hack, it is a small optimization, which allows not to add extra Motion node.

bandetto replied:
Alright. But does this make sense:
> Can the logic be moved to functions which added these nodes initially?

red1452 replied:
It is two different steps at generating path. There are two ways: we can add one more motion or just change locus at the existing motion. When we create the first motion (when change locus from segemntGeneral to signleQE) we do not know about the next changing locus to outerQuery. And we can not create motion with outerQuery locus.

bandetto reviewed (commented):
About PR description:
- How about adding some context on why nodes that have both a replicated table and a volatile function in their targetlist require special treatment?
- How about replacing "function ..." to "...()", just adding empty parentheses to signify it's a function instead of writing *function* each time? (e.g. "*Function cdbpath_create_motion_path*" -> "*cdbpath_create_motion_path()*")
- I suggest changing "*Such cases may be cased, when subquery contains ...*" -> "*Such cases may be caused by a subquery containing ...*".
- "*If there are Projection path and Motion nodes, Materialize will be added between these nodes, and also locus will be changed at these nodes.*" -> "*If there is a Projection path with Motion below, their locus will be updated, and a Materialize path will be added in between of them.*"
- "*For the other cases Materialize and Motion nodes will be added to the top of the path without modifying the other nodes.*" -> "*For other cases, Materialize and Motion paths will just be added on top  without modifying other nodes.*"

red1452 committed:
minor fix

red1452 committed:
Merge remote-tracking branch 'origin/adb-7.2.0' into ADBDEV-5518

bandetto commented on src/test/regress/expected/rpt.out:
```diff
@@ -1613,6 +1666,164 @@ drop table t1;
+(9 rows)
+
+-- ensure we make broadcast motion when volatile function in correlated subplan qual
+explain (costs off, verbose) select * from t1 where a in (select f(i) from t where i=a and f(i) > 0);
```
This is the only test that fails without the patch. Can you explain why they are needed in PR description?

red1452 replied:
All tests were taken from 6X. Another code changes can not be taken, because the sources of 7X were rewritten. But tests are useful.

bandetto replied:
> But tests are useful.

Okay. Can you explain in PR description, why?

red1452 replied:
I updated the description.

bandetto replied:
LGTM, but

> These test are useful

Why?

red1452 replied:
It checks correctness of planning queries with volatile functions and replicated tables.

bandetto commented on src/backend/cdb/cdbpath.c:
```diff
@@ -243,6 +242,48 @@ cdbpath_create_motion_path(PlannerInfo *root,
+			 */
+			else if (IsA(subpath, ProjectionPath))
+			{
+				ProjectionPath* projection_path = (ProjectionPath *) subpath;
```
Please indent your changes with pgindent.

red1452 replied:
Done.

bandetto commented on src/test/regress/expected/rpt.out:
```diff
@@ -1613,6 +1666,164 @@ drop table t1;
+(9 rows)
+
+-- ensure we make broadcast motion when volatile function in correlated subplan qual
+explain (costs off, verbose) select * from t1 where a in (select f(i) from t where i=a and f(i) > 0);
```
An incorrect version of this plan is included in the ticket, but for me, it just fails the planning altogether with an:
```
ERROR:  Passing parameters across motion is not supported. (cdbmutate.c:2051)
```
Why? Is this normal?

red1452 replied:
It is normal, because new check was added by [PR-1220](https://github.com/arenadata/gpdb/pull/1220).

red1452 committed:
run pgindent

red1452 committed:
Merge branch 'adb-7.2.0' into ADBDEV-5518

red1452 committed:
Merge branch 'adb-7.2.0' into ADBDEV-5518
